### PR TITLE
SDIT-1490 Add date matcher validation and remove body part matcher

### PR DESCRIPTION
**Pull request by mikehalmamoj**

### CODE REVIEW

File: hmpps-prisoner-search/src/test/kotlin/uk/gov/justice/digital/hmpps/prisonersearch/search/services/dto/AttributeSearchRequestTest.kt [hunk 1]
```diff
@@ -538,13 +538,16 @@ class AttributeSearchRequestTest(@Autowired val objectMapper: ObjectMapper) {
...
                      "condition": "IS",
                      "searchTerm": "John"
-                   }
-                 ],
-                 "bodyPartMatchers": [
+                   },
+                   {
+                     "attribute": "tattoos.bodyPart",
+                     "condition": "IS",
+                     "searchTerm": "shoulder"
```
Comment (mikehalmamoj) [inline hmpps-prisoner-search/src/test/kotlin/uk/gov/justice/digital/hmpps/prisonersearch/search/services/dto/AttributeSearchRequestTest.kt:545]: Removed specific body part matcher - it's just a text matcher on a nested object

File: hmpps-prisoner-search/src/test/kotlin/uk/gov/justice/digital/hmpps/prisonersearch/search/services/dto/AttributeSearchRequestTest.kt [hunk 2]
```diff
@@ -538,13 +538,16 @@ class AttributeSearchRequestTest(@Autowired val objectMapper: ObjectMapper) {
...
+                   },
                    {
-                     "attribute": "tattoo",
-                     "bodyPart": "shoulder",
-                     "comment": "dragon"
+                     "attribute": "tattoos.comment",
+                     "condition": "CONTAINS",
+                     "searchTerm": "dragon"
```
Comment (mikehalmamoj) [inline hmpps-prisoner-search/src/test/kotlin/uk/gov/justice/digital/hmpps/prisonersearch/search/services/dto/AttributeSearchRequestTest.kt:550]: Added contains to the text matcher

File: hmpps-prisoner-search/src/main/kotlin/uk/gov/justice/digital/hmpps/prisonersearch/search/services/AttributeSearchService.kt [hunk 3]
```diff
@@ -117,6 +127,28 @@ class AttributeSearchService {
...
+          throw AttributeSearchException("Attribute $attribute is not an date attribute")
+        }
+      }
+      ?: throw AttributeSearchException("Unknown attribute: $attribute")
+
+    if (minValue == null && maxValue == null) {
+      throw AttributeSearchException("Attribute $attribute must have at least 1 min or max value")
+    }
+
+    if (minValue != null && maxValue != null) {
```
Comment (petergphillips) [inline hmpps-prisoner-search/src/main/kotlin/uk/gov/justice/digital/hmpps/prisonersearch/search/services/AttributeSearchService.kt:143]: is it worth adding in future validation for min date and I guess too far in the past validation for max date?

File: hmpps-prisoner-search/src/test/kotlin/uk/gov/justice/digital/hmpps/prisonersearch/search/services/AttributeSearchServiceTest.kt [hunk 4]
```diff
@@ -388,5 +390,111 @@ class AttributeSearchServiceTest {
...
+              dateMatchers = listOf(
+                DateMatcher("firstName", minValue = today),
+              ),
+            ),
+          ),
+        )
+
+        assertThrows<AttributeSearchException> {
+          service.validate(request)
+        }.also {
```
Comment (petergphillips) [inline hmpps-prisoner-search/src/test/kotlin/uk/gov/justice/digital/hmpps/prisonersearch/search/services/AttributeSearchServiceTest.kt:415]: interesting, haven't seen this way of dealing with `assertThrows` before

**Review by petergphillips — APPROVED** [review]
LGTM

**Pull request merged**